### DISCUSSION TITLE
Many Unconnected Filler/Tap Cells in Design

### DISCUSSION THREAD

**Discussion by b224hisl** [The-OpenROAD-Project/OpenROAD-flow-scripts#2565] [Q&A]
I've encountered an issue. It's a very simple design that includes two GF180 SRAMs. When I run `analyze_power_grid`, I encounter many instances that are not properly connected to power. Do you know if this relates to my failure to correctly add the PDN with macros? If so, how can I ensure that all these instances are properly powered? 

In the GUI, I see that the instances that are reported not connected to power seem to have already made contact with VDD and VSS through `-followpins`. Could you explain what might be going on?

![QQ_1731762522689](https://github.com/user-attachments/assets/1b17576d-89cd-475a-b5f2-76c7f131c66e)
![QQ_1731762499320](https://github.com/user-attachments/assets/dbde12d5-abe7-445c-b121-f4c3c8966581)

This is my environment: (`/OpenRoad-flow-scripts/tools/OpenROAD# ./etc/Env.sh`)
[env.txt](https://github.com/user-attachments/files/17784797/env.txt)

To reproduce:
[final_report_2small_gf180_10_tsv_2024-11-16_21-11.tar.gz](https://github.com/user-attachments/files/17784754/final_report_2small_gf180_10_tsv_2024-11-16_21-11.tar.gz)

Here is the report of the unconnected instance ( `check_power_grid -net VDD -error_file error.rpt`):
[error.txt](https://github.com/user-attachments/files/17784790/error.txt)

Thanks very much!

**Comment by maliberty** [comment]
Can't reproduce
```
[ERROR ORD-0007] ./results/gf180/2small/10_tsv/6_1_fill.odb does not exist.
[ERROR GUI-0070] Error: final_report.tcl, 4 ORD-0007
```

btw, this would be better filed as an issue

**Reply by b224hisl** [reply]
> Can't reproduce
> 
> ```
> [ERROR ORD-0007] ./results/gf180/2small/10_tsv/6_1_fill.odb does not exist.
> [ERROR GUI-0070] Error: final_report.tcl, 4 ORD-0007
> ```
> 
> btw, this would be better filed as an issue

It should be good this time： 
[final_report_2small_gf180_test_2024-11-17_15-32.tar.gz](https://github.com/user-attachments/files/17789299/final_report_2small_gf180_test_2024-11-17_15-32.tar.gz)

**Comment by maliberty** [comment]
Btw, it is more natural to use
```
export ADDITIONAL_LEFS += $(MACRO_FOLDER)/gf180mcu_fd_ip_sram__sram64x8m8wm1.lef
export ADDITIONAL_LIBS += $(MACRO_FOLDER)/gf180mcu_fd_ip_sram__sram64x8m8wm1.lib
```
than
```
BLOCKS = gf180mcu_fd_ip_sram__sram64x8m8wm1
```
As I expect you are not building the sram in OR.

**Reply by maliberty** [reply]
Maybe you did as the VDD pin looks weirdly fragmented:
![image](https://github.com/user-attachments/assets/35077725-6cca-4b49-be44-0f912e7b1ec3)

**Reply by b224hisl** [reply]
I have included the SRAM block because I have modified the `Makefile`:
![QQ_1731825340926](https://github.com/user-attachments/assets/86fa4365-b3f9-49c9-b4b5-a1bc82f9c11a)
In the GUI, this SRAM block also appears to be fine. This is the  [Gf180 SRAM](https://github.com/google/globalfoundries-pdk-ip-gf180mcu_fd_ip_sram)
![QQ_1731825363835](https://github.com/user-attachments/assets/c9cb7e57-d98e-479e-aa28-274866b0c120)
I think the GF180 SRAM has its own power stripes, which I did not place myself.

**Comment by maliberty** [comment]
@gadfort could you help here?  Its a bit odd.  There are power pins on m1, eg
![image](https://github.com/user-attachments/assets/5ccbbc3d-89c8-4049-bcc1-8c85b54c994f)

The grid seems connected but everything is being flagged by psm:
![image](https://github.com/user-attachments/assets/b8d09bbe-377d-463c-a04d-1d38e8bb7a24)

Is there some assumption in psm about power coming in from above?

**Reply by maliberty** [reply]
Note it isn't just fill cells, every instance is marked disconnected.

**Reply by gadfort** [reply]
@maliberty just going a quick search through the odb, it looks like there are no `Via1` so yes, every instance is disconnected from the grid. It would take longer to go into a deeper dive.

**Reply by maliberty** [reply]
There are oddly few Metal4 stripes
![image](https://github.com/user-attachments/assets/2c64650f-ebd4-4b82-8569-be4a2bf7c033)

But when I look at them I see vias on all layers m1-m4:
![image](https://github.com/user-attachments/assets/74361855-37db-41d9-b453-f097c15ef5d3)

**Reply by gadfort** [reply]
@maliberty the database I'm looking at has no Via1

**Reply by maliberty** [reply]
Are you looking at 
It should be good this time：
[final_report_2small_gf180_test_2024-11-17_15-32.tar.gz](https://github.com/user-attachments/files/17789299/final_report_2small_gf180_test_2024-11-17_15-32.tar.gz)

**Reply by gadfort** [reply]
The power grid failed to generate (per the logs).
```
add_pdn_connect -grid {block} -layers {Metal1 Metal4} -max_columns {5} -ongrid {Metal2 Metal3 Metal4} -split_cuts {Metal3 0.128}
```

changing this to:
```
add_pdn_connect -grid {block} -layers {Metal1 Metal4}
add_pdn_connect -grid {block} -layers {Metal3 Metal4}
```

the grid generates a complete grid, except psm is reporting the memories as disconnected.
However, then attempted to dump the lefs I got:
```
[INFO ORD-0034] More than one lib exists, multiple files will be written.
[CRITICAL ODB-0438] Internal inconsistency: no table found for type 57
```

**Reply by gadfort** [reply]
Looking at the LEF, the VDD port is broken into "weakly" connected statements which PSM interprets are requiring the macro to be externally connected (which it is not)

**Reply by maliberty** [reply]
@b224hisl please include macro.tcl.

I turned off MACRO_PLACEMENT_TCL/BLOCKS, and added ADDITIONAL_LEFS/ADDITIONAL_LIBS in the config.mk.  I was able to run the flow and still see a version of the problem with somewhat different macro placement.

**Reply by maliberty** [reply]
You can see the inputs LEF at designs/gf180/2small/macro/gf180mcu_fd_ip_sram__sram64x8m8wm1.lef

**Reply by b224hisl** [reply]
> @b224hisl please include macro.tcl.
> 
> I turned off MACRO_PLACEMENT_TCL/BLOCKS, and added ADDITIONAL_LEFS/ADDITIONAL_LIBS in the config.mk. I was able to run the flow and still see a version of the problem with somewhat different macro placement.
Here is the `macro.tcl`

```
set block [ord::get_db_block]
set units [$block getDefUnits]
set vals [list 2 1 95 20 0 116]
lassign $vals rows cols offsetx offsety pitchx pitchy

for {set row 0} {$row < $rows} {incr row} {
    for {set col 0} {$col < $cols} {incr col} {
        set inst [$block findInst [format "genblk_%d_%d_sram_u" $row $col]]
        puts "row=$row,col=$col"
        set x [expr round( ($offsetx + (($pitchx) * $col)) * $units )]
        set y [expr round( ($offsety + (($pitchy + 232) * $row)) * $units )]
        $inst setOrigin $x $y
        puts "x=$x"
        puts "y=$y"
        $inst setPlacementStatus FIRM
    }
}
```

**Reply by b224hisl** [reply]
> The grid seems connected but everything is being flagged by psm:



> The power grid failed to generate (per the logs).
> 
> ```
> add_pdn_connect -grid {block} -layers {Metal1 Metal4} -max_columns {5} -ongrid {Metal2 Metal3 Metal4} -split_cuts {Metal3 0.128}
> ```
> 
> changing this to:
> 
> ```
> add_pdn_connect -grid {block} -layers {Metal1 Metal4}
> add_pdn_connect -grid {block} -layers {Metal3 Metal4}
> ```
> 
> the grid generates a complete grid,

By the way, I would like to ask how I can determine that the PDN grid is fully/completly generated. As shown in this example, I add power stripes in Metal1 and Metal4, why the `add_pdn_connect` should be applied in {Metal1 Metal4} and  {Metal3 Metal4}? I have reviewed the documentation for `add_pdn_connect`, but I am still unclear on this point.

**Comment by maliberty** [comment]
@gadfort would a power-ring make this easier?

**Reply by gadfort** [reply]
What do you mean? A power ring around the memory, sure, but you would need enough m4 straps to intersect all the memory power pins.

**Reply by maliberty** [reply]
Yes around the RAM.  Wouldn't you just need to connect the straps to the ring and not all the power pins?  Separately you would need stubs to tie the ram pins to the ring.

**Reply by maliberty** [reply]
There looks to be an internal power ring in the gds
![image](https://github.com/user-attachments/assets/7f8fed59-387c-4f70-96b8-9da6092b4c06)

No idea why these are weak connects

**Reply by gadfort** [reply]
It's how they are expressed in the LEF file (I cannot see the GDS from within PSM).
If the LEF was modified to make 1 PORT with lots of PINs then PSM would not complain since LEF specifies that as strongly connected internally.

**Reply by gadfort** [reply]
a ring would not be required, but the LEF being wrong doesn't help.

**Reply by maliberty** [reply]
I don't know what hope there is for a corrected LEF.  The blockages are pretty terrible as well.

**Comment by maliberty** [comment]
I pulled the gds from https://github.com/google/globalfoundries-pdk-ip-gf180mcu_fd_ip_sram

**Reply by b224hisl** [reply]
OK thank you very much~
